### PR TITLE
Remove metadata-related code after its move to -deprecated

### PR DESCRIPTION
**Pull request by mih**
This changeset is part of the transition from the 2nd-gen metadata implementation to then gen3 code provided by the -metalad extension.

This removal has side-effect on other code, not directly related to metadata. This following list can be considered a TODO for further maintenance work:

Now unused code:

- `datalad/support/json_py.py:load_xzstream()` : `load_xzstream` and `dump_xzstream` _removed in 82631286443cddde9051bb32f00f4dfc878d237b_
- `datalad/interface/common_opts.py:reporton_opt`: _removed in 1c2607230dc65f7f4995dc0031810b5ab208db5a_
- `datalad/utils.py:all_same()`: _removed in 5e50038ef4f4182d22b19448b60d54bdc6b49721_
- `datalad/utils.py:as_unicode()`: _removed in 13514bad9c43fa73fdba39b5b3a5e151483bf351_
- `datalad/utils.py:unicode_srctypes` made obsolete by removal of `as_unicode()`: _removed in 13514bad9c43fa73fdba39b5b3a5e151483bf351 (same as above)_

Now barely used (technical debt?):

- `nosave_opt`: Residual use in `addurls()` and `download_url()`
- `datalad/support/path.py:split_ext()`: Residual use in a single spot in `addurls()` <- https://github.com/datalad/datalad/issues/4056
- `datalad/utils.py:path_startswith()`: 2 pieces of code still use this outside the tests
- `datalad/utils.py:get_suggestions_msg()`: CLI parser and `addurls()` make the total of 2 calls made to this helper
- `datalad/utils.py:shortened_repr()`: one call in `get()` and one in `foreach_dataset()`

Issues to file:

- [x] Investigate whether the dependency on `simplejson` is still needed, `simplejson` can be removed by PR datalad/datalad#7035 
- [x] **Won't fix now** `datalad/interface/utils.py:discover_dataset_trace_to_targets()` is now an internal helper of `save()`
- [x] Clean uses the removed `SEARCH_INDEX_DOTGITDIR`
- [x] Fix doc-building
- [x] Move code to deprecated (datalad-deprecated#62 (https://github.com/datalad/datalad-deprecated/issues/62))

Related TODOs:

- close or move all remaining metadata-related issues https://github.com/datalad/datalad/labels/metadata
- https://github.com/datalad/datalad-deprecated/issues/74
- https://github.com/datalad/datalad-deprecated/issues/75
- https://github.com/datalad/datalad-deprecated/issues/76
- https://github.com/datalad/datalad-deprecated/issues/77
- Update `_datalad_build_support` and remove `BuildSchema` and related code -- maybe that was only in -core?

Closes datalad/datalad#7012

- [x] remove temporary commit testing against a PR in deprecated rather than the released version
- [x] double-check the list above and changelog

### CODE REVIEW

**Comment by christian-monch** [comment]
I think the dependency on `simplejson` can be removed in favor of the built-in `json`-module. The only use outside of `json_py.py` is in `annexrepo.py` and there it can just be replaced with the corresponding `json`-calls. The comment seems to indicate that `simplejson` was used to be able to log errors, but that is done `json`-based code as well.

**Comment by christian-monch** [comment]
Should we move this out of draft and review it?

The extension tests can best be fixed after this PR is released. 


Besides removing metadata-related code and fixing tests, the following has been done

- All "now unused code" removed (see description for the respective commits)
- Investigation whether the dependency on `simplejson` is still required has been performed. `simplejson` is no longer required and replaced with `json` (mainly done in: 66c2a63a2a38992b3a670a151689f4a5de99d8b4, minor fixes in additional commits)
- Removed SEARCH_INDEX_DOTGITDIR (in: 7289576ed554cd3444b11f24af064bad1ca389da)


The following has NOT been tackled yet in this PR:
- Remove "Now barely used (technical debt?)" code
- No action on "datalad/interface/utils.py:discover_dataset_trace_to_targets() is now an internal helper of save()"
- Docs need fixing

File: datalad/consts.py [hunk 1]
```diff
@@ -35,8 +35,6 @@
 ANNEX_TEMP_DIR = join('.git', 'annex', 'tmp')
 ANNEX_TRANSFER_DIR = join('.git', 'annex', 'transfer')
 
-SEARCH_INDEX_DOTGITDIR = join('datalad', 'search_index')
-
```
Comment (mih) [inline datalad/consts.py:50]: I would leave it in (no cost) and add a comment that it is deprecated/ignored.

File: datalad/local/clean.py [hunk 2]
```diff
@@ -133,8 +132,6 @@ def __call__(*,
                  "temporary annex", FILES_PLURAL),
                 (Path(ANNEX_TRANSFER_DIR), "annex-transfer",
                  "annex temporary transfer", DIRS_PLURAL),
-                (gitdir / Path(SEARCH_INDEX_DOTGITDIR), 'search-index',
-                 "metadata search index", FILES_PLURAL),
```
Comment (mih) [inline datalad/local/clean.py:137]: I think this was my reason to keep the const.

**Review by mih — COMMENTED** [review]
Some comments

File: datalad/local/clean.py (hunk 2 above)
Comment (christian-monch) [reply datalad/local/clean.py]: Kept it in the code

File: datalad/consts.py (hunk 1 above)
Comment (christian-monch) [reply datalad/consts.py]: Thx. Suggestion applied

**Comment by yarikoptic** [comment]
appveyor mac fail: `FAILED ../datalad/support/tests/test_parallel.py::test_stalling - AssertionError: Future has not finished in 10x time` . in 0.17.4 (https://github.com/datalad/datalad/pull/6995) we boosted from 5 to 10... odd that it is still not enough, might need to boost to some 100 and if that ever fails -- there must be a real hard to reproduce issue

**Comment by christian-monch** [comment]
> appveyor mac fail: `FAILED ../datalad/support/tests/test_parallel.py::test_stalling - AssertionError: Future has not finished in 10x time` . in 0.17.4 (#6995) we boosted from 5 to 10... odd that it is still not enough, might need to boost to some 100 and if that ever fails -- there must be a real hard to reproduce issue

I am not able to reproduce it locally on `darwin 19.6.0`. Since it seems unrelated to the metadata move, it would suggest to solve this in a dedicated issue

**Comment by bpoldrack** [comment]
Ok, so still not ready. Apparently the move to deprecated is still incomplete. datalad-deprecated is trying to import annotate_path stuff from core, that has moved to deprecated.

**Comment by bpoldrack** [comment]
Remaining failures are currently happening everywhere and are clearly unrelated. Looking into this.

However, extension tests now pass. I'm gonna go ahead and merge this finally.

**Comment by yarikoptic-gitmate** [comment]
PR released in [`0.18.0`](https://github.com/datalad/datalad/releases/tag/0.18.0)

**Pull request merged**